src/hashcash.cr

```diff
@@ -1,6 +1,141 @@
 # TODO: Write documentation for `Hashcash`
+
+require "base64"
+require "digest/sha1"
+
 module Hashcash
-  VERSION = "0.1.0"
+  # Hashcash.generate("resource")
+  # => 1:20:201206222555:resource::pOWgc88+uDuefr/o:MTMxNzg2MA==
+  # OR can customise hashcash defaults
+  # Hashcash.generate("hello", version: 1, bits: 16, date: Time.utc, ext: "goodbyye")
+  # => 1:16:201206222403:hello:goodbyye:kfwaGRadlD3ddc9G:MTMxMzY5NQ==
+  def self.generate(
+    resource : String,
+    version = 1,
+    bits = 20,
+    date = Time.utc,
+    ext = ""
+  )
+    Hashcash::Stamp.new(resource, version, bits, date, ext).generate
+  end
+
+  # Hashcash.verify?("1:20:201206222555:resource::pOWgc88+uDuefr/o:MTMxNzg2MA==", "resource")
+  # => true
+  # Hashcash.verify?("invalid_string", "resource")
+  # => false
+  def self.verify?(hashcash_stamp : String, resource : String, time_window = 2.days.ago..2.days.from_now, bits = 20) : Bool
+    Hashcash::Stamp.new(resource).valid?(hashcash_stamp, resource, time_window, bits)
+  end
+
+  class Stamp
+    STAMP_VERSION = 1
+    getter version, bits, date, resource, ext, stamp_string
+
+    def initialize(
+      @resource : String,
+      @version = STAMP_VERSION,
+      @bits = 20,
+      @date = Time.utc,
+      @ext = "",
+      @stamp_string = ""
+    )
+    end
+
+    def generate : String
```

Comment:
I'd recommend splitting this so that there is a `#to_s` that forms a string, and a seperate `#generate` method that performs the hashing needed to set `counter`.